pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -391,7 +393,12 @@ class Timestamp(_Timestamp):
             # User passed tzinfo instead of tz; avoid silently ignoring
             tz, tzinfo = tzinfo, None
 
-        if isinstance(ts_input, str):
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        if (isinstance(ts_input, Timestamp) and not
+                any(arg is not None for arg in _non_ts_attributes)):
```

Comment:
The same as with _date_attributes.

Reply:
This makes sense.  One question since we're focusing on performance: does it make a difference if you write out the verbose `and foo is None and bar is None and baz is None...`?  i have a suspicion that cython doesnt optimize this list comprehension

Reply:
@jbrockmendel Let's test it then.
The current implementation:
```
>>> ts = pd.Timestamp("2019-01-01")
>>> timeit.timeit(lambda: pd.Timestamp(ts), number=10000000)
5.860465700000002
```

And now with this implementation:
```
        if (isinstance(ts_input, Timestamp) and freq is None and
                tz is None and
                unit is None and
                year is None and
                month is None and
                day is None and
                hour is None and
                minute is None and
                second is None and
                microsecond is None and
                nanosecond is None and
                tzinfo is None):
            return ts_input
```

```
>>> ts = pd.Timestamp("2019-01-01")
>>> timeit.timeit(lambda: pd.Timestamp(ts), number=10000000)
4.43885800000001
```
You are right. Cython probably builds the entire list before starting boolean checks, so we incur on overhead.